### PR TITLE
DEPR: Index.reindex with duplicate index

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Discussed on this week's call.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3888,6 +3888,15 @@ def reindex(
             if self.equals(target):
                 indexer = None
             else:
+                if not self.is_unique:
```
Comment (jreback) [inline pandas/core/indexes/base.py:3891]: don't you need to do this on L3912? e.g. raise if those conditions are tru first. *also* don't we *know* that its unique at this point?

**Review by jreback — CHANGES_REQUESTED** [review]
comment

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:3891]: > don't you need to do this on L3912? e.g. raise if those conditions are tru first

sure, will move this down

> also don't we know that its unique at this point?

nope

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -3902,6 +3903,15 @@ def reindex(
                         )
                     indexer, _ = self.get_indexer_non_unique(target)
 
+                if not self.is_unique:
```
Comment (jreback) [inline pandas/core/indexes/base.py:3906]: isn't this the same check as L3892?

Comment (jbrockmendel) [reply pandas/core/indexes/base.py:3906]: no, self.is_unique is _slightly_ different from self._index_as_unique

**Review by jreback — APPROVED** [review]

**Pull request merged**